analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java

```diff
@@ -0,0 +1,196 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.fasten.analyzer.qualityanalyzer;
+
+import eu.fasten.analyzer.qualityanalyzer.data.QAConstants;
+
+import eu.fasten.core.data.Constants;
+import eu.fasten.core.plugins.KafkaPlugin;
+import eu.fasten.core.plugins.DBConnector;
+
+import org.jooq.exception.DataAccessException;
+import org.json.JSONObject;
+import org.pf4j.Extension;
+import org.pf4j.Plugin;
+import org.pf4j.PluginWrapper;
+
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
+
+import java.sql.BatchUpdateException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.List;
+import java.util.Collections;
+
+import org.jooq.DSLContext;
+
+
+public class QualityAnalyzerPlugin extends Plugin {
+
+    public QualityAnalyzerPlugin(PluginWrapper wrapper) {
+        super(wrapper);
+    }
+
+    @Extension
+    public static class QualityAnalyzer implements KafkaPlugin, DBConnector {
+
+        private final Logger logger = LoggerFactory.getLogger(QualityAnalyzer.class.getName());
+        private String consumerTopic = "fasten.RapidPlugin.out";
+        private static MetadataUtils utils = null;
+        private Throwable pluginError = null;
+
+        @Override
+        public void setDBConnection(Map<String, DSLContext> dslContexts) {
+            QualityAnalyzer.utils = new MetadataUtils(dslContexts);
+        }
+
+        @Override
+        public Optional<List<String>> consumeTopic() {
+            return Optional.of(Collections.singletonList(consumerTopic));
+        }
+
+        @Override
+        public void setTopic(String topicName) {
+            this.consumerTopic = topicName;
+        }
+
+        @Override
+        public void consume(String kafkaMessage) {
+
+            logger.info("Consumed: " + kafkaMessage);
+
+            var jsonRecord = new JSONObject(kafkaMessage);
+
+            String forge = null;
+
+            if (jsonRecord.has("payload")) {
+                forge = jsonRecord
+                        .getJSONObject("payload")
+                        .getString("forge".replaceAll("[\\n\\t ]", ""));
+            }
+
+            logger.info("forge = " + forge);
+
+            if(forge == null) {
+                logger.error("Could not extract forge from the message");
+                setPluginError(new RuntimeException("Could not extract forge from the message"));
+                return;
+            }
+
+            boolean processedRecord = false;
+            int transactionRestartCount = 0;
+            boolean restartTransaction = false;
+
+            Long recordId = null;
+
+            do {
```

Comment:
If you are trying to use transactions, then you are not actually starting a transaction. Here is the link that nicely explains how JOOQ transaction management works https://www.jooq.org/doc/3.12/manual/sql-execution/transaction-management/
If you don't want transactions but only retries, then you don't need that empty catch block. It is needed for proper error management in transactions (you can take a closer look at how this works in the `MetadataDatabasePlugin` class - transaction starts with lambda function, there is an inner catch block which decides whether to restart transaction or not and it always throws caught exception back from the transaction lambda to stop the transaction, then outer empty catch block catches everything but doesn't do anything because everything was already done in the inner catch block, and then transaction is restarted if needed).

Reply:
OK, the latest version is there. I moved transactions to `MetadataUtils`, in order (hopefully) to get a more readable code, and hopefully I handle all Exceptions. I still need to throw `JsonParsingException` instead of `IllegalStateException` where applicable.

Reply:
Yes, the transaction management looks correct now. Thanks!